### PR TITLE
New syntax/puppet.vim

### PR DESCRIPTION
**Pull request by thallgren**
Features:
- Recognizes functions from core and stdlib
- Recognizes resource types from core
- Recognizes Pcore Types
- Recognizes expressions found in string interpolations.
- Handles @(DOC) and @("DOC") heredoc structures

The origin of this file was the syntax/ruby.vim found at:

https://github.com/vim-ruby/vim-ruby

It was used as the base because it provided good support for interpolation
expressions and heredoc.

### CODE REVIEW

**Comment by reidmv** [comment]
Getting syntax support for Puppet 4 features is a big win IMO. Seems like maybe next steps would be for someone who knows what they're doing to compare the functionality of this proposed syntax script with the current script and see if there are any gaps. Any takers?

**Comment by rski** [comment]
@thallgren @reidmv I suggest pinging people on irc about this, #voxpupuli on freenode. This seems like a good patch but I no longer use vim so I wouldn't like to review this.

File: .gitignore
Comment (rski) [inline .gitignore:1]: This is...odd for a vim plugin

File: syntax/puppet.vim
Comment (rski) [inline syntax/puppet.vim:7]: I'd like to keep these, with voxpupuli as the maintainer and url pointing to this repo

**Review by rski — CHANGES_REQUESTED** [review]
There are some odd things here and there maintenance-wise, but I really won't bother with looking at the functional aspects of the patch.

File: syntax/puppet.vim
Comment (rski) [reply syntax/puppet.vim:7]: All of these fields, not just the ones that got selected.

File: .gitignore
Comment (thallgren) [reply .gitignore:1]: Does it matter though? It's the editor of choice for me and having it in there is rather convenient.

Comment (rski) [reply .gitignore:1]: yeah, sure, just odd.

File: syntax/puppet.vim
Comment (thallgren) [reply syntax/puppet.vim:7]: I didn't include them because they had no content. Some comments regarding them:
- URL: Sure, it could point to this repo.
- Last Change: Can be determined by looking at the URL. If present, it should probably be updated automatically in each commit or it will diverge. What scheme should be used for this?
- Version: What versioning scheme do you want? A SHA? A Semver? Regardless, why not use a version file that denotes the version currently in the repository. It's not unlikely that future improvements span more than this particular file.
- Filename: What's the point in including a file name in a named file?

**Comment by rski** [comment]
It looks good and the features it provides are promising and I really appreciate the work put in this. I _could_ merge this and let people complain if it breaks, but I'd never fix it and that would be a not so community friendly move.

File: syntax/puppet.vim
Comment (rski) [reply syntax/puppet.vim:7]: On version and filename, fair enough. Last change is a convention for Vim files, even Vim's own files follow it.

Comment (thallgren) [reply syntax/puppet.vim:7]: There's only one other file in this repo that uses a 'Last change'. It's the indent/puppet.vim that contains:

" Last Change: 2009 Aug 19

which is obviously quite incorrect. I can add it, but are you sure you want this field added with the burden of maintaining it? Doesn't seem to be something that anyone cared much about in the past.

**Review by rski — APPROVED** [review]

File: syntax/puppet.vim
Comment (rski) [reply syntax/puppet.vim:7]: Ok yup, you're right. It is a convention and it would be nice to have it, but nothing is better than wrong.

**Review by reidmv — APPROVED** [review]
I've been using this version of the syntax file for more than a week and have had no problems with it. I like it a lot better than the previous one because it supports all the Puppet 4 syntax in my files, which the previous syntax file did not. 👍

**Comment by rski** [comment]
@reidmv thanks for the feedback. Have you encountered anything fishy?

**Comment by reidmv** [comment]
As I said, no problems. Now that I'm scratching around for something, anything, to fix, there's a minor highlighting bug I found around escape sequences and interpolation in double-quotes:

syntax

That's the biggest, fishiest issue I could find with it. @thallgren ^^

**Comment by rski** [comment]
I think if that is fixed, I'll merge it.

**Comment by rski** [comment]
Ok it works now :-)
@thallgren can you squash the commits? I'm being pedantic, but I'll merge it after.

**Comment by thallgren** [comment]
Squashed, per request.

**Comment by rski** [comment]
@thallgren perfect. Thank you for the contribution, this is some good work you've done \o/

**Pull request merged**
